### PR TITLE
build-system: Introduce CSS Cache 🚀

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Final step for [#33011](https://github.com/ampproject/amphtml/issues/33011)

The slowest part of AMP's dev server startup is transforming CSS via `postcss`. This PR takes the infrastructure we built for [caching babel transforms](https://github.com/ampproject/amphtml/pull/33314) which was surprisingly generic, and reuses it for caching our css transforms.

_results_:
- warm cache css build reduced from 2.7s --> 100ms.
- total lazy server startup time for a warm cache is now 1s.
- extension building should be sped up as well, although these were already pretty fast so much less interesting.

**implementation notes**

- Renamed `BabelTransformCache` to `TransformCache`, and moved it from  `esbuild-babel.js` to its own file.
- Specified cache location as `.css-cache` (corollary to `.babel-cache`) as well as added it to the `gulp clean` command and the `.gitignore` file.
- A combination of `package.json`, `postcss` options, and each individual `filehash` are used for the cache key.

### CODE REVIEW

File: build-system/common/esbuild-babel.js [hunk 1]
```diff
@@ -15,77 +15,19 @@
...
 const babel = require('@babel/core');
-const crypto = require('crypto');
-const fs = require('fs-extra');
 const path = require('path');
+const {TransformCache, batchedRead, md5} = require('./transform-cache');
 
 /**
  * Directory where the babel filecache lives.
  */
 const CACHE_DIR = path.resolve(__dirname, '..', '..', '.babel-cache');
```
Comment (rsimha) [inline build-system/common/esbuild-babel.js:24]: I wonder if the directory location should now be a detail of implementation of `TransformCache`. E.g. callers pass in just the name of the cache, and the class decides where it lives, and provides a way to get the full path from the cache name.

**Review by rsimha — COMMENTED** [review]
Nice PR! I've made a couple of initial comments. Would love to get Justin's take on this as well.

File: build-system/tasks/css/jsify-css.js [hunk 2]
```diff
@@ -35,6 +41,8 @@ const browsersList = {
   ],
 };
 
+// See http://cssnano.co/optimisations/ for full list.
```
[the remark this answers is not in the document]
Comment (samouri) [reply build-system/tasks/css/jsify-css.js:44]: Good spot! Fixed

File: .gitignore [hunk 3]
```diff
@@ -6,6 +6,7 @@ build/
 .amp-dep-check
 extensions/**/dist/**
 c
+.css-cache/
```
[the remark this answers is not in the document]
Comment (samouri) [reply .gitignore:9]: Done

File: build-system/common/esbuild-babel.js (hunk 1 above)
Comment (samouri) [reply build-system/common/esbuild-babel.js:24]: I lean slightly towards agreeing

File: build-system/common/transform-cache.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  set(hash, transformPromise) {
+    if (this.transformMap.has(hash)) {
+      throw new Error(
+        `Read race occured. Attempting to transform a file twice.`
+      );
+    }
+
+    this.transformMap.set(hash, transformPromise);
+    transformPromise.then((contents) => {
+      fs.outputFile(path.join(this.cacheDir, hash), contents);
```
Comment (jridgewell) [inline build-system/common/transform-cache.js:67]: Nit: might be nice to include a file extension?

**Review by jridgewell — APPROVED** [review]

File: build-system/common/transform-cache.js (hunk 4 above)
Comment (samouri) [reply build-system/common/transform-cache.js:67]: WDYM? Previously we had `[hash].json`. Could make it `[hash].cache`?

Comment (jridgewell) [reply build-system/common/transform-cache.js:67]: We're writing the raw files now, so `.js` or `.css`?

**Review by rsimha — APPROVED** [review]

File: build-system/common/esbuild-babel.js (hunk 1 above)
Comment (samouri) [reply build-system/common/esbuild-babel.js:24]: Discussed via Chat and decided to kick this down the road a bit

Comment (rsimha) [reply build-system/common/esbuild-babel.js:24]: This was done in #33834.

**Pull request merged**